### PR TITLE
add cardinality to shape element descriptor

### PR DESCRIPTION
**Pull request by fmoor**
This change alters the binary protocol by adding information about the
maximum and minimum number of elements in a shape's field to the object
type descriptor.

working towards https://github.com/edgedb/edgedb-go/issues/135

### CODE REVIEW

**Review by tailhook — COMMENTED** [review]
Looks like you have to bump `DUMP_VER_MAX`

File: docs/internals/protocol/typedesc.rst [hunk 1]
```diff
@@ -73,7 +73,8 @@ Object Shape Descriptor
         //   1 << 0: the field is implicit
         //   1 << 1: the field is a link property
         //   1 << 2: the field is a link
-        uint8           flags;
+        //   7 << 3: Cardinality
```
Comment (elprans) [inline docs/internals/protocol/typedesc.rst:76]: Perhaps it'd be cleaner to split the upper and lower cardinality into two binary flags instead?

Comment (elprans) [reply docs/internals/protocol/typedesc.rst:76]: @1st1 ^

Comment (fmoor) [reply docs/internals/protocol/typedesc.rst]: changed

**Review by 1st1 — APPROVED** [review]

**Comment by 1st1** [comment]
Good work!

File: edb/server/compiler/sertypes.py [hunk 2]
```diff
@@ -65,11 +66,36 @@
...
+
+        raise RuntimeError("unreachable")
+
+
 class TypeSerializer:
 
     EDGE_POINTER_IS_IMPLICIT = 1 << 0
     EDGE_POINTER_IS_LINKPROP = 1 << 1
     EDGE_POINTER_IS_LINK = 1 << 2
+    EDGE_POINTER_IS_REQUIRED = 1 << 3
```
Comment (elprans) [inline edb/server/compiler/sertypes.py:97]: This seems to be dead code.

Comment (fmoor) [reply edb/server/compiler/sertypes.py:97]: :facepalm:

Comment (fmoor) [reply edb/server/compiler/sertypes.py:97]: https://github.com/edgedb/edgedb/pull/2612

**Pull request merged**